### PR TITLE
Mesh: fixes Mesh.to / DomainMesh.to casting integer cells to a float dtype

### PR DESCRIPTION
**Pull request by peterdsharpe**
# PhysicsNeMo Pull Request

## Description

`mesh.to(torch.float64)` (and `DomainMesh.to`) raised 'TypeError: cells must have an int-like dtype': the generated tensorclass `.to` casts EVERY leaf tensor, including the integer cells, which then fails `Mesh.__post_init__`. So the dtype conversion documented in the README was broken for any mesh with cells.

Override `.to` so that a floating/complex dtype is applied only to floating tensors -- integer cells (and any integer data) are preserved. Device-only moves (and non-float dtypes) delegate unchanged to the generated tensorclass `.to`, so device metadata, non_blocking, etc. behave exactly as before; only a floating dtype cast takes the cells-safe path. The target dtype/device is resolved by probing with a zero-length slice of points (no data copy). `DomainMesh.to` mirrors this, delegating per-mesh through the fixed `Mesh.to` and casting `global_data`.

Adds regression tests (Mesh and DomainMesh) covering float64 cast, integer-cells/data preservation, cache cast, point clouds, and device moves.

## Checklist

- [ ] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.
- [ ] If I am implementing a new model or modifying any existing model, I have followed the [Models Implementation Coding Standards](https://github.com/NVIDIA/physicsnemo/wiki/Coding-standards-for-models-implementation).

## Review Process

**All PRs are reviewed by the PhysicsNeMo team before merging.**

Depending on which files are changed, GitHub may automatically assign a maintainer for review.

We are also testing AI-based code review tools (e.g., Greptile), which may add automated comments with a confidence score.
This score reflects the AI’s assessment of merge readiness and is **not** a qualitative judgment of your work, nor is
it an indication that the PR will be accepted / rejected.

AI-generated feedback should be reviewed critically for usefulness.
You are not required to respond to every AI comment, but they are intended to help both authors and reviewers.
Please react to Greptile comments with 👍 or 👎 to provide feedback on their accuracy.

### CODE REVIEW

File: physicsnemo/mesh/mesh.py [hunk 1]
```diff
@@ -3060,3 +3060,48 @@ def _mesh_repr(self) -> str:
...
+    if cast_dtype is None:
+        # Device move and/or non-float dtype: the generated tensorclass ``to`` is
+        # correct (it never turns the integer cells into a float dtype) and preserves
+        # the device metadata shown in repr.
+        return _tensorclass_mesh_to(self, *args, **kwargs)
+
+    # Floating/complex dtype cast: move to the target device with the generated ``to``
+    # (cells-safe, sets device metadata), then cast only the floating leaves so the
+    # integer cells (and any integer data) are never cast to a float dtype.
+    moved = _tensorclass_mesh_to(self, device=probe.device)
```
[the remark this answers is not in the document]
Comment (negin513) [reply physicsnemo/mesh/mesh.py]: 👍 correct fix for `non_blocking`, but recommend the generalized version to also
  cover `memory_format` and other kwargs...

Comment (negin513) [inline physicsnemo/mesh/mesh.py]:
```suggestion
    extra = {k: v for k, v in kwargs.items() if k not in 
  ("dtype", "device")}
   moved = _tensorclass_mesh_to(self, device=probe.device, 
  **extra)
```

Something like this for passing the `kwargs`?

File: test/mesh/mesh/test_init_validation.py [hunk 2]
```diff
@@ -356,3 +356,35 @@ def test_valid_combinations(
         assert mesh.n_cells == n_cells
         assert mesh.n_spatial_dims == n_spatial_dims
         assert mesh.n_manifold_dims == n_manifold_dims
+
+
+def test_to_float_dtype_preserves_integer_cells_and_data():
```
Comment (negin513) [inline test/mesh/mesh/test_init_validation.py:362]: This looks great.

File: physicsnemo/mesh/domain_mesh.py [hunk 3]
```diff
@@ -1161,3 +1161,32 @@ def _domain_mesh_repr(self: DomainMesh) -> str:
...
+# metadata-preserving); only a floating dtype cast takes the per-mesh path through the
+# (cells-safe) ``Mesh.to`` via ``apply_to_meshes``, with ``global_data`` cast too.
+def _domain_mesh_to(self, *args: Any, **kwargs: Any) -> "DomainMesh":
+    # Zero-length probe: resolves the requested dtype/device without copying data.
+    probe = self.interior.points[:0].to(*args, **kwargs)
+    cast_dtype = None
+    if probe.dtype != self.interior.points.dtype and (
+        probe.dtype.is_floating_point or probe.dtype.is_complex
+    ):
+        cast_dtype = probe.dtype
```
[the remark this answers is not in the document]
Comment (negin513) [reply physicsnemo/mesh/domain_mesh.py]: I think greptile is correct here as well... The logic here is broken similar to the mesh.py for cases where the DomainMesh is already at the requested float dtype.

**Review by negin513 — CHANGES_REQUESTED** [review]
Nice fix! One thing to sort out before merge: the `probe.dtype != points.dtype` guard re-triggers the original TypeError when casting to the dtype  mesh already had, e.g. float64 --> float64 (same in mesh.py and domain_mesh.py).
Maybe adding a test for this edge case?

File: physicsnemo/mesh/domain_mesh.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (peterdsharpe) [reply physicsnemo/mesh/domain_mesh.py]: Yep :D Fixing it!

**Comment by peterdsharpe** [comment]
/ok to test https://github.com/NVIDIA/physicsnemo/pull/1716/commits/76f37efc3f6a73442c2a9c6f424c6f87d0168335

File: physicsnemo/mesh/mesh.py (hunk 1 above)
Comment (peterdsharpe) [reply physicsnemo/mesh/mesh.py]: Added a variant of this (for `dtype`, not `device`)

File: physicsnemo/mesh/mesh.py [hunk 4]
```diff
@@ -3060,3 +3060,48 @@ def _mesh_repr(self) -> str:
...
+def _mesh_to(self, *args: Any, **kwargs: Any) -> "Mesh":
+    # Probe with a zero-length slice of the (always-floating) points tensor to
+    # resolve the requested dtype/device, reusing torch's own ``.to`` overload
+    # parsing without copying any data.
+    probe = self.points[:0].to(*args, **kwargs)
+    cast_dtype = None
+    if probe.dtype != self.points.dtype and (
+        probe.dtype.is_floating_point or probe.dtype.is_complex
+    ):
+        cast_dtype = probe.dtype
```
[the remark this answers is not in the document]
Comment (peterdsharpe) [reply physicsnemo/mesh/mesh.py]: Fixed + added test

**Review by negin513 — APPROVED** [review]

**Comment by peterdsharpe** [comment]
/ok to test https://github.com/NVIDIA/physicsnemo/pull/1716/commits/631dc9c5faa9de2ad278d859a94fb0de447a32a3

**Comment by peterdsharpe** [comment]
/ok to test https://github.com/NVIDIA/physicsnemo/pull/1716/commits/111f4aae716a6f17e0f71ac30c208229b0975a53

**Comment by peterdsharpe** [comment]
/ok to test https://github.com/NVIDIA/physicsnemo/pull/1716/commits/5c100b5a08dd7d0b5100db9b812a06d8472557b1

**Comment by peterdsharpe** [comment]
/ok to test https://github.com/NVIDIA/physicsnemo/pull/1716/commits/182b26dd5dc95ee2b437c2c64fda91855702acfa
